File: dataStructures/Stack/bestDigits.js
// Write a function that takes a positive integer represented as a string number and an integer numDigits. Remove numDigtis from the string so that the number represent by the string is a large as possible afterwards.
// Note that the order of the remaining digits cannot be changed. You can assume numDigits will always be less than the length of number and greater than or equal to 0.

function bestDigits(number, numDigits) {
  const stack = [];

  // The important concept here is that, in order to keep the largest number possible, we want to remove the smallest numbers first, until we have run out of numbers to remove.
  for (let numChar of number) {
    // you don't need to convert the numChar to a number because you can compare a single digit string with greater than or less than and its the same (does not apply to numbers with multiple digits)
    while (
      stack.length > 0 &&
      numDigits > 0 &&
      numChar > stack[stack.length - 1]
    ) {
      stack.pop();
      numDigits--;
    }
    stack.push(numChar);
  }

  // After traversing the string, there may still be some numbers we can chop off, so this is needed in case, for the final number
  while (numDigits > 0) {
    stack.pop();
    numDigits--;
  }

  let result = stack.join("");
  return result;
}

File: dataStructures/Stack/bestDigits.py
# Algoexpert solution:
def bestDigits(number, numDigits):
    stack = []

    for digit in number:
        while numDigits > 0 and len(stack) > 0 and digit > stack[len(stack) - 1]:
            numDigits -= 1
            stack.pop()

        stack.append(digit)

    while numDigits > 0:
        numDigits -= 1
        stack.pop()

    return "".join(stack)


# my first solution - only passed 16/18 tests
def bestDigits(number, numDigits):
    length = len(number)
    i = 1
    result_list = [number[0]]

    while i < length:
        j = i

        while j < length and len(result_list) > 0 and int(number[j]) >= int(result_list[-1]) and numDigits > 0:
            result_list.pop()
            numDigits -= 1
            j += 1

        result_list.append(number[i])
        i += 1

    while numDigits > 0:
        result_list.pop()
        numDigits -= 1

    result = "".join(result_list)
    print(result)
    return result


num = "10000000002"
digit = 9

print(bestDigits(num, digit))
